### PR TITLE
Handle tax rule for multishop

### LINKED ISSUE #35926: [CQRS] Add/Update multistore
Handle shop association in the CQRS commands/queries
- force array of shopIds in the create command
- partial update in the update command
- return list of associated shops in the query

Update behat tests accordingly

### PR DESCRIPTION
**Pull request by tleon**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Handle ShopConstraint and get tax rule out of carrier command and tests
| Type?             | improvement
| Category?         | BO 
| BC breaks?        | yes
| Deprecations?     | yes
| How to test?      | CI ok
| UI Tests          | Please run UI tests and paste here the link to the run. [Read this page to know why and how to use this tool](https://devdocs.prestashop-project.org/8/contribute/contribution-guidelines/ui-tests/).
| Fixed issue or discussion?     | Fixes #35926 
| Sponsor company   | PrestaShop SA

### CODE REVIEW

File: src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php
Comment (boherm) [inline src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php]: Or:
```suggestion
        if (!$command->getShopConstraint()->forAllShops()) {
```
To avoid creating of new ShopConstraint.

File: src/Adapter/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandler.php
Comment (boherm) [inline src/Adapter/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandler.php]: Same here.

File: src/Core/Domain/Carrier/Command/SetCarrierTaxRuleGroupCommand.php
Comment (boherm) [inline src/Core/Domain/Carrier/Command/SetCarrierTaxRuleGroupCommand.php]: It's preferable to use scalar values into the constructor here. So you must pass an integer for `$carrierId` instead of `CarrierId`, but in the constructor you create a new CarrierId with this integer and fill a `private CarrierId $carrierId` class property with this new object.

```suggestion
    private CarrierId $carrierId;
    
    public function __construct(
        int $carrierId,
        private int $carrierTaxRuleGroupId,
        private ShopConstraint $shopConstraint
    ) {
        $this->carrierId = new CarrierId($carrierId);
    }
```

File: tests/Integration/Behaviour/Features/Context/Domain/Carrier/CarrierTaxRuleGroupFeatureContext.php
Comment (boherm) [inline tests/Integration/Behaviour/Features/Context/Domain/Carrier/CarrierTaxRuleGroupFeatureContext.php]: I'm not blocking for this as we haven't optimise the duplication system for carriers but we need to reflect that system in step sentence as: 
```suggestion
     * @When I set tax rule for carrier :reference called :newReference with specified properties:
```

File: src/Adapter/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandler.php
Comment (boherm) [inline src/Adapter/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandler.php]: We need to duplicate the carrier before with `CarrierRepository::updateInNewVersion` and return the new CarrierId.

File: tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_tax_rule_group_management.feature
Comment (boherm) [inline tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_tax_rule_group_management.feature:33]: Maybe, we can add a test to check when `taxRuleGroup` is invalid 🤔

**Review by boherm — CHANGES_REQUESTED** [review]
It's look promising! Just some few things 😉
We can discuss about it if you need.

File: src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php
Comment (jolelievre) [inline src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php]: I believe @boherm intriduce a dedicated exception or at least an error code But this can be homogenized in another PR

File: src/Adapter/Carrier/Validate/CarrierValidator.php
Comment (jolelievre) [inline src/Adapter/Carrier/Validate/CarrierValidator.php]: Seems like this method is used nowhere

File: src/Core/Domain/Carrier/Command/SetCarrierTaxRuleGroupCommand.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/Command/SetCarrierTaxRuleGroupCommand.php]:
```suggestion
    public function getCarrierTaxRuleGroupId(): TaxRulesGroupId
```
You can use this existing VO

File: src/Core/Domain/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandlerInterface.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandlerInterface.php]:
```suggestion
    public function handle(SetCarrierTaxRuleGroupCommand $command): CarrierId;
```

Comment (jolelievre) [reply src/Core/Domain/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandlerInterface.php]: You should return the new carrier ID when it's been duplicated (which later will be the same potentially once this part has been optimized)

File: tests/Integration/Behaviour/Features/Context/Domain/Carrier/CarrierFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/Carrier/CarrierFeatureContext.php:295]: I think the assertion part can remain in this step, especially since the tax rules validation in the other context is still based on the same Query, it would be simpler to keep all the assertions in one place

File: tests/Integration/Behaviour/Features/Context/Domain/Carrier/CarrierTaxRuleGroupFeatureContext.php
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/Carrier/CarrierTaxRuleGroupFeatureContext.php]: Indeed the same as what was done for the edit step Although the wording is improbable but for now we should keep the same for consistency

File: tests/Integration/Behaviour/behat.yml
Comment (jolelievre) [inline tests/Integration/Behaviour/behat.yml]:
```suggestion
            - Tests\Integration\Behaviour\Features\Context\Domain\Carrier\CarrierFeatureContext
            - Tests\Integration\Behaviour\Features\Context\Domain\Carrier\CarrierTaxRuleGroupFeatureContext
            - Tests\Integration\Behaviour\Features\Context\Domain\CustomerFeatureContext
```

**Review by jolelievre — CHANGES_REQUESTED** [review]

**Review by jolelievre — DISMISSED** [review]

**Review by jolelievre — APPROVED** [review]

File: src/Adapter/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandler.php
Comment (boherm) [inline src/Adapter/Carrier/CommandHandler/SetCarrierTaxRuleGroupHandler.php:51]: This way will change soon ;)

**Review by boherm — APPROVED** [review]
LGTM, thanks @tleon!

**Comment by jolelievre** [comment]
Thanks @tleon

**Comment by jolelievre** [comment]
Rock and roll 🤘

**Comment by ps-jarvis** [comment]
PR merged, well done!

Message to @PrestaShop/committers: do not forget to milestone it before the merge.

**Pull request merged**
